### ISSUE TITLE
TLC reports temporal violation even in the presence of safety violation

### ISSUE THREAD

**Issue by jeenuv** [tlaplus/tlaplus#699]
I think that safety violations should be reported before temporal violations.

I've an example checked in [here](https://bitbucket.org/jeenuv/test_spec/src/potential-bug/AudioChannel.tla). I've put in a [`Print`](https://bitbucket.org/jeenuv/test_spec/src/fe910454b764b6f847595a7484fee178ed47c14b/AudioChannel.tla#lines-121) from TLC module where the expression is occasionally printed as `FALSE`, but that safety property is never reported as violated. Instead, as mentioned, a temporal property violation (stuttering) is reported.

The [fix for the violation is also present](https://bitbucket.org/jeenuv/test_spec/src/fe910454b764b6f847595a7484fee178ed47c14b/AudioChannel.tla#lines-44:47) in the file; if that's uncommented, both the safety and temporal violations go away.

Command to run:
```
tlc -deadlock AudioChannel.tla
```

Tested on both 1.7.1 and 1.8.0.

**Comment by lemmy** [comment]
Please run TLC with '-lncheck final' (on the command-line), or check "Verify temporal properties upon termination only" on the Toolbox's "TLC options" page if you want to defer liveness checking until *after* the state-graph has been constructed,

**Comment by jeenuv** [comment]
Tried:

```
tlc -lncheck final -deadlock AudioChannel.tla
```

This doesn't seem to have any effect, though (tested on both 1.7.1 and 1.8.0). I also notice that the option is not listed with `-help`, but TLC does recognize it.

**Comment by lemmy** [comment]
You've stated `NoOvertaking` as a temporal property (`[]`) and, thus, have to check it as a `PROPERTY`.  Remove the box operator and change it to `INVARIANT` in the config file to make the `Print` output on stdout look as expected. 

By the way, the `Print` operator is not a good tool to debug a spec.

**Comment by jeenuv** [comment]
Yup, I confirm your suggestion works. With this fix, I also note that I don't need to specify the `-lncheck final` option.

It's interesting to know the TLC's choices on safety and liveness checking. I'd like to however confirm I understand this correctly: writing `INVARIANT NoOvertaking` and checking `[]NoOvertaking` via `PROPERTY` are semantically same, aren't they? That is both assert that, on conforming behaviours, `NoOvertaking` is maintained at all times? Only difference is that TLC employs two different algorithms in the two cases? (I'm going by page 225 of Specifying Systems here).

Presumably the effect of `-lncheck final` per your first suggestion was to force TLC to choose a different algorithm than usual? I'm also curious as to why it's not documented.

> By the way, the Print operator is not a good tool to debug a spec.

I'm aware! My native habitat is the command line, where I don't suppose I have much choice with TLC. There's feature on trace that I've noticed, but haven't explored it much.

**Comment by lemmy** [comment]
Yes, `INVARIANT NoOvertaking` and `PROPERTY []NoOvertaking` are semantically the same in the logic of TLA+.

`-lncheck final` defers the general algorithm (safety & liveness) until after the algorithm that checks safety is done.  Without `-lncheckd final`, TLC periodically runs the general algorithm on the incomplete graph.
